crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move

```diff
@@ -412,6 +415,181 @@ module sui::kiosk {
+    /// extension completely, it keeps it from performing any protected actions.
+    /// The storage is still available to the extension (until it's removed).
+    public fun disable_extension<Ext: drop>(
+        self: &mut Kiosk,
+        cap: &KioskOwnerCap,
+    ) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        extension_mut<Ext>(self).is_enabled = false;
+    }
+
+    /// Re-enable permissions for the extension. Can only be performed by the
+    /// owner. The extension can start performing protected actions again.
+    public fun re_enable_extension<Ext: drop>(
+        self: &mut Kiosk,
+        cap: &KioskOwnerCap,
+    ) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        extension_mut<Ext>(self).is_enabled = true;
+    }
+
+    /// Limit the extension to a set of types. Can only be performed by the
+    /// owner. The extension can only perform protected actions on the types
+    /// specified in the list.
+    public fun limit_extension_for_types<Ext: drop>(
+        self: &mut Kiosk,
+        cap: &KioskOwnerCap,
+        types: vector<TypeName>,
+    ) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        extension_mut<Ext>(self).types = types;
+    }
+
+    /// Get immutable access to the extension storage. Can only be performed by
+    /// the extension as long as the extension is installed.
+    public fun ext_storage<Ext: drop>(
+        _ext: Ext, self: &mut Kiosk
+    ): &mut Bag {
+        &mut extension_mut<Ext>(self).storage
+    }
+
+    /// Get mutable access to the extension storage. Can only be performed by
+    /// the extension as long as the extension is installed. Removing
+    /// permissions does not prevent the extension from accessing the storage.
+    ///
+    /// Potentially dangerous: extension developer can keep data in a Bag
+    /// therefore never really allowing the KioskOwner to remove the extension.
+    /// However, it is the case with any other solution (1) and this way we
+    /// prevent intentional extension freeze when the owner wants to ruin a
+    /// trade (2) - eg locking extension while an auction is in progress.
+    ///
+    /// Extensions should be crafted carefully, and the KioskOwner should be
+    /// aware of the risks.
+    public fun ext_storage_mut<Ext: drop>(
+        _ext: Ext, self: &mut Kiosk
+    ): &mut Bag {
+        &mut extension_mut<Ext>(self).storage
+    }
+
+    /// Protected action: place an item into the Kiosk. Can be performed by an
+    /// authorized extension. The extension must have the `place` permission
+    /// and the type of the item must be in the list of allowed types.
+    public fun ext_place<Ext: drop, T: key + store>(
+        _ext: Ext, self: &mut Kiosk, item: T
+    ) {
+        let ext = extension<Ext>(self);
+
+        assert!(ext.is_enabled, EExtensionDisabled);
+        assert!(ext.permissions & 1 != 0, EExtensionNotAllowed);
```

Comment:
I think it would be easier to read if we expose a `has_place` or `assert_place_permission` function (and similar for the other permissions) rather than doing the bit arithmetic here

Reply:
It's a very careful implementation approach because the final set of permissions is still under discussion. For example: should `place` be a separate permission with `lock`? While `place` is innocent, `lock` may lead to an unwanted asset being stuck forever in a Kiosk. At the same time locking and placing mechanics behave symmetrically in trades.